### PR TITLE
Infiniband parcelport 

### LINKED ISSUE #1401: Need an efficient infiniband parcelport
Recent developments on parcelport redesign make it worthwhile to reimplement the infiniband verbs based parcelport.

### PR DESCRIPTION
**Pull request by biddisco**
Still got some problems with multiple nodes connecting to each other, but want to start the process of testing and getting ready for a merge into master.

This branch should fix #1401 and #840

JAN - 2017 : I have updated the PR with new connection code and a bunch of other improvements. - it should be 'good enough' for production use now, at least on smallish numbers of nodes, it has been tested up to 128 successfully.

I will now address the original review comments and fix style/inspect report fails.

### CODE REVIEW

File: plugins/parcelport/verbs/header.hpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    struct header
+    {
+        static constexpr unsigned int header_block_size = sizeof(detail::header_block);
+        static constexpr unsigned int data_size_        = SIZE - header_block_size;
+        static const     unsigned int chunk_flag        = 0x01;
+        static const     unsigned int message_flag      = 0x02;
+        //
+    private:
+        detail::header_block           message_header;
+        boost::array<char, data_size_> data_;
```
Comment (hkaiser) [inline plugins/parcelport/verbs/header.hpp]: What's the difference between boost::array and std::array? Would it be possible to use the latter?

File: plugins/parcelport/verbs/readers_writers_mutex.hpp [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+ * field cycles and the same ticket is taken by another thread, we could grant
+ * a lock to two separate threads at the same time, and bad things happen: two
+ * writer threads or a reader thread and a writer thread would run in parallel,
+ * and lock waiters could be skipped if the unlocks race. This is unlikely, it
+ * only happens if a lock request is blocked by 64K other requests. The fix is
+ * to grow the lock structure fields, but the largest atomic instruction we have
+ * is 8 bytes, the structure has no room to grow.
+ */
+// from http://locklessinc.com/articles/locks/
+
```
Comment (hkaiser) [inline plugins/parcelport/verbs/readers_writers_mutex.hpp]: Please protect this file from being used on non-gcc compilers.
Also, should this file go into `hpx/lcos/local`?

File: plugins/parcelport/verbs/scheduler.hpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
+//  Copyright (c) 2015-2016 John Biddiscombe
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
```
Comment (hkaiser) [inline plugins/parcelport/verbs/scheduler.hpp]: Is this file still needed? Wasn't that superseded by one of the executor classes we added?

File: plugins/parcelport/verbs/unordered_map.hpp [hunk 4]
```diff
@@ -0,0 +1,354 @@
+////////////////////////////////////////////////////////////////////////////////
+//  Copyright (c) 2016 John Biddiscombe
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+////////////////////////////////////////////////////////////////////////////////
+
```
Comment (hkaiser) [inline plugins/parcelport/verbs/unordered_map.hpp]: I trust this will move to the directory where the other concurrent containers live (will live).

File: cmake/HPX_SetupBoost.cmake [hunk 5]
```diff
@@ -21,6 +21,11 @@ set(Boost_ADDITIONAL_VERSIONS
     "1.58.0" "1.58"
     "1.57.0" "1.57")
 
+set(HPX_BOOST_LOG_LIBS "")
+if(HPX_PARCELPORT_VERBS_WITH_LOGGING OR HPX_PARCELPORT_VERBS_WITH_DEV_MODE)
+  set(HPX_BOOST_LOG_LIBS log log_setup)
+endif()
```
Comment (biddisco) [inline cmake/HPX_SetupBoost.cmake]: The majority of the rdma code was originally in its own mini library and was using boost log. The logging code is only activated when enabled via cmake options and I would not expect anyone other than myself or another developer interested in the verbs parcelport to activate it.
I agree that we should have just one logging options, but this has proven to be my #1 mean of diagnosing problems and leaving it in for the time being is not going to cost anyone. In the long term we should migrate to a single logging infrastructure - but I'd like to postpone this until this is proven to be production stable.

File: hpx/util/memory_chunk_pool_allocator.hpp [hunk 6]
```diff
@@ -106,6 +107,7 @@ namespace hpx { namespace util { namespace detail
         }
 
         Pool * memory_pool_;
+        bool   disable_deallocate;
```
Comment (biddisco) [inline hpx/util/memory_chunk_pool_allocator.hpp]: I believe this is left over from some earlier experimentation - the verbs PP passes pinned memory blocks to the parcel decode functions where they are freed, but there was a need to disable the deallocation so that the blocks could be reused. I suspect this is now obsolete and I will investigate furtuer and address this.

File: plugins/parcelport/verbs/header.hpp (hunk 1 above)
Comment (biddisco) [inline plugins/parcelport/verbs/header.hpp]: I will replace all boost::array instances with std::array, most of the code is cleaned up, but a few of these are inevitably still present.

File: plugins/parcelport/verbs/readers_writers_mutex.hpp (hunk 2 above)
Comment (biddisco) [inline plugins/parcelport/verbs/readers_writers_mutex.hpp]: My plan was to migtrate this to lcos and the concurrent map to hpx::concurrent, but only after this is merged to master and work on hpx::concurrent is more mature.

File: plugins/parcelport/verbs/scheduler.hpp (hunk 3 above)
Comment (biddisco) [inline plugins/parcelport/verbs/scheduler.hpp]: Yes, this scheduler is only enabled by cmake under user control and comes with a warning (see cmakelists). Work on custom schedulers is ongoing and I'd like to leave this code in place until I'm ready to replace it. Please feel free to create an issue to address custom parcelport exectutors as a placeholder to get rid of this.

File: plugins/parcelport/verbs/unordered_map.hpp (hunk 4 above)
Comment (biddisco) [inline plugins/parcelport/verbs/unordered_map.hpp]: As per comment above, the plan is to move this to hpx::concurrent once things settle down. this unordered map code is now well tested and has proven to be very useful with the readers/writers lock guarding it.

File: plugins/parcelport/verbs/pinned_memory_vector.hpp [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+                throw std::runtime_error(
+                    "pinned_memory_vector should never be resized once an "
+                    "allocation has been assigned");
+            }
+            m_region_ = m_alloc_->allocate_region(s);
+            m_array_ = static_cast<T*>(m_region_->get_address());
+        }
+
+    private:
+        pinned_memory_vector(pinned_memory_vector<T> const & other);
```
Comment (sithhell) [inline plugins/parcelport/verbs/pinned_memory_vector.hpp]: You should use `HPX_MOVABLE_ONLY(pinned_memory_vector)` here.

File: plugins/parcelport/verbs/rdma/rdma_controller.hpp [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+            int(std::pair<uint32_t, uint64_t>, verbs_endpoint_ptr)>
+        ConnectionFunction;
+        typedef std::function<int(verbs_endpoint_ptr client)>
+        DisconnectionFunction;
+
+        // Set a callback which will be called immediately after
+        // RDMA_CM_EVENT_CONNECT_REQUEST has been received. This can be useful to
+        // prevent two connections being established at the same time.
+        void setConnectRequestFunction(ConnectRequestFunction f) {
+            this->request_connect_function_ = f;
```
Comment (sithhell) [inline plugins/parcelport/verbs/rdma/rdma_controller.hpp]: `f` should be moved here.

File: plugins/parcelport/verbs/rdma/rdma_controller.hpp [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+        verbs_endpoint* getClient(uint32_t qp) {
+            return clients_[qp].get();
+        }
+
+        rdma_memory_pool_ptr getMemoryPool() {
+            return memory_pool_;
+        }
+
+        template<typename Function>
+        void for_each_client(Function lambda)
```
Comment (sithhell) [inline plugins/parcelport/verbs/rdma/rdma_controller.hpp]: Use a forwarding reference here:
```
void for_each_client(Function&& lambda)
{
    std::for_each(clients_.begin(), clients_.end(), std::forward<Function>(lambda));
}
```

File: plugins/parcelport/verbs/readers_writers_mutex.hpp [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+        //
+        void lock()
+        {
+            // memory ordering barrier
+            rwl_barrier_();
+
+            uint16_t val = atomic_xadd(&ticket.s.next, 1);
+            while (val != ticket.s.writers) {
+// std::cout << "rw lock sleep write " << val << " " << ticket.s.next << " \n";
+                hpx::util::detail::yield_k(4, nullptr);
```
Comment (sithhell) [inline plugins/parcelport/verbs/readers_writers_mutex.hpp]: Why not use an exponential back off here?

File: plugins/parcelport/verbs/readers_writers_mutex.hpp [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+        //
+        void lock_shared()
+        {
+            // memory ordering rwl_barrier_
+            rwl_barrier_();
+
+            uint16_t val = atomic_xadd(&ticket.s.next, 1);
+            while (val != ticket.s.readers) {
+// std::cout << "rw lock sleep read " << val << " " << ticket.s.readers << "\n";
+                hpx::util::detail::yield_k(0, nullptr);
```
Comment (sithhell) [inline plugins/parcelport/verbs/readers_writers_mutex.hpp]: Same here, couldn't we use an exponential backoff here as well?

File: src/util/command_line_handling.cpp [hunk 12]
```diff
@@ -667,8 +667,7 @@ namespace hpx { namespace util
             }
         }
 
-        if ((vm.count("hpx:connect") || mode_ == hpx::runtime_mode_connect) &&
-            hpx_host == "127.0.0.1")
+        if (hpx_host == "127.0.0.1" || hpx_host == "localhost")
```
Comment (sithhell) [inline src/util/command_line_handling.cpp]: Is this code here still needed? I think that's what's causing the hang on circleci.

File: hpx/runtime/parcelset/parcelport_connection.hpp [hunk 13]
```diff
@@ -92,6 +93,10 @@ namespace hpx { namespace parcelset {
         parcelport_connection(typename BufferType::allocator_type const & alloc)
           : buffer_(alloc)
         {}
+
+        parcelport_connection(typename BufferType::allocator_type * alloc)
+          : buffer_(std::move(buffer_type(alloc)),alloc)
```
Comment (biddisco) [inline hpx/runtime/parcelset/parcelport_connection.hpp]: I'll check this and fix it.

File: plugins/parcelport/mpi/CMakeLists.txt [hunk 14]
```diff
@@ -47,8 +47,8 @@ if(HPX_WITH_PARCELPORT_MPI)
     endif()
 
     set(_mpi_libraries)
-    if(MPI_CXX_LIBRARIES)
-      set(_mpi_libraries ${_mpi_libraries} ${MPI_CXX_LIBRARIES})
+    if(MPI_C_LIBRARIES)
```
Comment (biddisco) [inline plugins/parcelport/mpi/CMakeLists.txt]: It may be the case that with some older cmake versions that you had problems, however - this is the correct behaviour and CXX libs should not be needed. I suggest you give this a try and if you run into problems on some machines then I will revert this.

File: plugins/parcelport/verbs/connection_handler.hpp [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+namespace verbs
+{
+        uint32_t get_verbs_device_address(const char *devicename, const char *iface,
+            char *hostname)
+        {
+          FUNC_START_DEBUG_MSG
+          // Find the address of the I/O link device.
+          verbs_device_ptr linkDevice;
+          try {
+            linkDevice = verbs_device_ptr(new verbs_device(devicename, iface));
```
Comment (biddisco) [inline plugins/parcelport/verbs/connection_handler.hpp]: good question. I will check if I reuse the device and make it a static instantiation if I don't need a smart_pointer.

File: plugins/parcelport/verbs/connection_handler.hpp [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+          }
+          catch (rdma_error& e) {
+            LOG_ERROR_MSG("error opening InfiniBand device: " << e.what());
```
Comment (biddisco) [inline plugins/parcelport/verbs/connection_handler.hpp]: I'm not sure what to do about errors in the parcelport in general. The error handling is at best 'inconsistent' - in some places I throw, and in others I just log an error and return a fail. 
Here's we should probably rethrow the error and allow the app to terminate, or be handled above. 
An #issue to make the error handling consistent would be a good idea.

File: plugins/parcelport/verbs/connection_handler.hpp [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+              << linkDevice->get_interface_name());
+
+          std::stringstream temp;
+          in_addr_t addr = linkDevice->get_address();
+          temp
+            << (int)((uint8_t*)&addr)[0] << "."
+            << (int)((uint8_t*)&addr)[1] << "."
+            << (int)((uint8_t*)&addr)[2] << "."
+            << (int)((uint8_t*)&addr)[3] << std::ends;
+          strcpy(hostname, temp.str().c_str());
```
Comment (biddisco) [inline plugins/parcelport/verbs/connection_handler.hpp]: No idea. I'll check this

File: plugins/parcelport/verbs/pinned_memory_vector.hpp (hunk 7 above)
Comment (biddisco) [inline plugins/parcelport/verbs/pinned_memory_vector.hpp]: ok. I'll fix it

File: plugins/parcelport/verbs/rdma/rdma_controller.hpp (hunk 8 above)
Comment (biddisco) [inline plugins/parcelport/verbs/rdma/rdma_controller.hpp]: I'll fix this too

File: plugins/parcelport/verbs/rdma/rdma_controller.hpp (hunk 9 above)
Comment (biddisco) [inline plugins/parcelport/verbs/rdma/rdma_controller.hpp]: and fix this

File: plugins/parcelport/verbs/readers_writers_mutex.hpp (hunk 10 above)
Comment (biddisco) [inline plugins/parcelport/verbs/readers_writers_mutex.hpp]: The readers writers lock is used in some critical code where latency could be an issue. I wanted a very fast check and recheck, because this lock should not be held for long periods, so I didn't want a possible suspend, just a true spin wait on the lock. To enable the 'exponential backoff' do I just replace the 4 with a larger number (32 is a key one if I recall correctly), or should I use a different construct?

File: plugins/parcelport/verbs/readers_writers_mutex.hpp (hunk 11 above)
Comment (biddisco) [inline plugins/parcelport/verbs/readers_writers_mutex.hpp]: same reply as above

File: src/util/command_line_handling.cpp (hunk 12 above)
Comment (biddisco) [inline src/util/command_line_handling.cpp]: I'd better check that. we have some issues with the hostnames that this resolves, but it may introduce new problems. srun and manual start behave differently and for one I needed IB ip addresses, for the other I needed TCP ip addresses. It's a bit of a mess and I'm not sure what to do for the general case. I'll test more and see if hermione/rostam can be fixed without breaking my local usage.

File: plugins/parcelport/verbs/readers_writers_mutex.hpp (hunk 10 above)
Comment (sithhell) [inline plugins/parcelport/verbs/readers_writers_mutex.hpp]: the construct to use for exponential backoff would be something like this:
```
std::size_t k = 0;
while (check) { hpx::util::detail::yield_k(k, nullptr); }
```

this will do a spin most of the time and only yield for special numbers of k.

Comment (biddisco) [inline plugins/parcelport/verbs/readers_writers_mutex.hpp]: should `k` be incremented in there?

File: plugins/parcelport/verbs/readers_writers_mutex.hpp (hunk 11 above)
Comment (hkaiser) [inline plugins/parcelport/verbs/readers_writers_mutex.hpp]: You should use `yield_k as:

    for (std::size_t k = 0; val != ticket.s.readers; ++k)
    {
        hpx::detail::yield_k(k);
    }

**Comment by biddisco** [comment]
This passed the checks. If it's ever going to be merged - now's the time! (I think most of the review comments were addressed).
